src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala

```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
       }
     }
 
+  def operator = "||" | "+" | "-" | "*" | "/" | "&" | "::"
+  val aggregationFunctions = Set("MAX", "SUM", "MIN", "ARRAY_ACCUM", "ARRAY_AGG")
+
+  // expressions
+  def expr : Parser[Expr] =
+    ( lexpr ~ operator ~ expr ^^ { case (lhs ~ op ~ rhs) => BinaryOpExpr(lhs, op, rhs) }
+    | lexpr
+    )
+
+  def lexpr : Parser[Expr] =
+    ( functionName ~ "(" ~ rep1sep(expr, ",") ~ ")" ^^ { 
+        case (name ~ _ ~ args ~ _) => FuncExpr(name, args, (aggregationFunctions contains name))
+      }
+    | constant ^^ { ConstExpr(_) }
+    | variableName ^^ { VarExpr(_) }
+    | "(" ~> expr <~ ")" 
+    )
 
   // TODO support aggregate function syntax somehow
-  def cqHead = relationName ~ "(" ~ repsep(variableName, ",") ~ ")" ^^ {
+  def cqHead = relationName ~ "(" ~ rep1sep(expr, ",") ~ ")" ^^ {
       case (r ~ "(" ~ variableUses ~ ")") =>
         Atom(r, variableUses.zipWithIndex map {
-          case(name,i) => Variable(name, r, i)
+          case (e,i) => ColumnExpr(e, r, i)
         })
     }
 
-  // TODO add conditional expressions for where clause
-  def cqConditionalExpr = failure("No conditional expression supported yet")
-  def cqBodyAtom: Parser[Atom] =
-    ( relationName ~ "(" ~ repsep(variableName, ",") ~ ")" ^^ {
-        case (r ~ "(" ~ variableBindings ~ ")") =>
-          Atom(r, variableBindings.zipWithIndex map {
-            case(name,i) => Variable(name, r, i)
-          })
+  // conditional expressions
+  def compareOperator = "LIKE" | ">" | "<" | ">=" | "<=" | "!=" | "=" | "IS" | "IS NOT"
+  def cond : Parser[Cond] = 
+    ( acond ~ (";") ~ cond ^^ { case (lhs ~ op ~ rhs) =>
+        BinaryOpCond(lhs, LogicOperator.OR, rhs)
+      }
+    | acond
+    )
+  def acond : Parser[Cond] = 
+    ( lcond ~ (",") ~ acond ^^ { case (lhs ~ op ~ rhs) => BinaryOpCond(lhs, LogicOperator.AND, rhs) }
+    | lcond
+    )
+  // ! has higher priority...
+  def lcond : Parser[Cond] = 
+    ( "!" ~> bcond ^^ { NegationCond(_) }
+    | bcond
+    )
+  def bcond : Parser[Cond] = 
+    ( expr ~ compareOperator ~ expr ^^ { case (lhs ~ op ~ rhs) =>
+        ComparisonCond(lhs, op, rhs)
       }
-    | cqConditionalExpr
+    | "[" ~> cond <~ "]"
     )
+
+  def cqBodyAtom: Parser[Atom] =
+    relationName ~ "(" ~ repsep(expr, ",") ~ ")" ^^ {
+      case (r ~ "(" ~ variableBindings ~ ")") =>
+        Atom(r, variableBindings.zipWithIndex map {
+        case (e,i) => ColumnExpr(e, r, i)
+      })
+    }
   def cqBody: Parser[List[Atom]] = rep1sep(cqBodyAtom, ",")
+
+  def cqBodyWithCondition = cqBody ~ ("," ~> cond).? ^^ {
+    case (b ~ c) => BodyWithCondition(b, c)
+  }
+
   def conjunctiveQuery : Parser[ConjunctiveQuery] =
-    cqHead ~ ":-" ~ rep1sep(cqBody, ";") ^^ {
-      case (headatom ~ ":-" ~ disjunctiveBodies) =>
-        ConjunctiveQuery(headatom, disjunctiveBodies)
-    }
+    cqHead ~ opt("*") ~ ":-" ~ rep1sep(cqBodyWithCondition, ";") ^^ {
```

Comment:
I wish we used a better syntax/symbol for `isDistinct`.

How about using a `?` after the head or name of the head?  It's like saying the head is interested in whether there exists a tuple or not.  I think we can remove the `?` from the variable relation schema to reduce confusion and require the type annotations `:: Boolean` for variable relations.

Postfixing head with `!` may also be a good choice since it [denotes uniqueness in math](https://en.wikipedia.org/wiki/Exclamation_mark#Mathematics).  However, we're kind of reserving it for negation, so I'm a little reluctant to push harder on this direction.